### PR TITLE
Fix uv custom index URLs omitted from model `requirements.txt`

### LINKED ISSUE #22881: [BUG] uv produces unusable requirements when pyproject.toml declares a custom index
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Issues Policy acknowledgement

- [x] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Databricks

### MLflow version

Client: 3.11.1
Tracking server: 3.11.1

### System information

OS Platform and Distribution: Databricks Runtime / Linux (Ubuntu)
Python version: 3.12.3

### Describe the problem

When `mlflow.flavor.log_model` is called in a project that has a pyproject.toml and uv.lock (either auto-detected or provided via the uv_project_path parameter), and the project declares a custom index, `uv export` produces a requirements list that omits the index information. 

The resulting requirements file is not installable downstream(for example in model serving), since the packages sourced from the custom index cannot be resolved.

While this appears to be intended behavior(see [issue in uv repo](https://github.com/astral-sh/uv/issues/13572)) of uv export itself, it effectively prevents using uv as a replacement for the existing dependency-inference path whenever a custom index is involved.

### Tracking information

MLflow version: 3.11.1
Python: 3.12.3

### Code to reproduce issue

pyproject.toml

```toml
[project]
name = "mlflow-uv-index-repro"
version = "0.1.0"
requires-python = ">=3.12"
dependencies = [
    "scikit-learn>=1.3",
    "my-internal-utils>=0.2",   # lives on a private index
]

[[tool.uv.index]]
name = "internal"
url = "https://pypi.internal.example.com/simple"
default = true

[tool.uv.sources]
my-internal-utils = { index = "internal" }
```

failure_demonstation.py
```python
import mlflow
from sklearn.linear_model import LinearRegression
from sklearn.datasets import make_regression

X, y = make_regression(n_samples=50, n_features=4)
model = LinearRegression().fit(X, y)

# uv auto-detection kicks in because uv.lock + pyproject.toml exist in CWD
with mlflow.start_run() as run:
    info = mlflow.sklearn.log_model(model, name="model")
```

### Stack trace

An attempt to pip install this requirements would cause:

```
ERROR: Could not find a version that satisfies the requirement my-internal-utils==0.2.7
ERROR: No matching distribution found for my-internal-utils==0.2.7
```

### Other info / logs

```
REPLACE_ME
```

### What component(s) does this bug affect?

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [ ] `area/docs`: MLflow documentation pages

### PR DESCRIPTION
**Pull request by serena-ruan**
### Related Issues/PRs

Closes #22881

### What changes are proposed in this pull request?

When a uv project declares a custom package index via `[[tool.uv.index]]`, `uv export` intentionally omits the index URL from its output (by design — see [uv#13572](https://github.com/astral-sh/uv/issues/13572)). MLflow then logs the incomplete requirements, causing `pip install` to fail at model serving time because packages sourced from the private index can't be resolved.

**Fix:** After `export_uv_requirements()` succeeds, call the existing `extract_index_urls_from_uv_lock()` helper to find all non-PyPI registry URLs in `uv.lock`, and prepend an `--extra-index-url <url>` line for each one to the returned requirements list.

The `--extra-index-url` entries survive the downstream `_deduplicate_requirements` and `warn_dependency_requirement_mismatches` pipelines because both functions already handle non-standard requirement strings gracefully.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests

Two new unit tests in `tests/utils/test_uv_utils.py`:
- `test_infer_pip_requirements_prepends_extra_index_urls_for_private_indexes` — verifies `--extra-index-url` is added when a uv.lock references a non-PyPI registry
- `test_infer_pip_requirements_no_extra_index_urls_when_no_private_indexes` — verifies no `--extra-index-url` lines are added for PyPI-only projects

Manual end-to-end repro: created a real uv project whose lock file resolved packages through `https://pypi-proxy.dev.databricks.com/simple/`. Logged a pyfunc model; the generated `requirements.txt` now correctly contains:
```
--extra-index-url https://pypi-proxy.dev.databricks.com/simple/
numpy==2.2.6; python_full_version < "3.11"
numpy==2.4.4; python_full_version >= "3.11"
```

### Does this PR require documentation update?

- [x] No.

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.

#### Is this a user-facing change?

- [x] Yes. When logging a model from a uv project that uses a custom package index, MLflow now includes `--extra-index-url` entries in the generated `requirements.txt` so that private packages can be resolved at model serving time.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

This pull request and its description were written by Claude.

### CODE REVIEW

**Review by harupy — APPROVED** [review]
LGTM
